File: static/tasks.js
// Timer initialization
var timerIdle;

// ===========================================================================
// Generate <p> HTML for the Task structure
// IN: task struct {	ID : string, Text : string, Section : string, Status : string, Icon : string, Timestamp : string }
// ===========================================================================
function htmlPTask(task) {
	var tooltips = {'created': statusCreated, 'moved': statusMoved, 'canceled': statusCanceled, 'done': statusDone};
	var p = '<p id="' + task.Id + '" class="' + task.Status + '" data-tooltip="' + tooltips[task.Status] + '" data-timestamp="'+task.Timestamp+'">'
	if (task.Icon) {
		p = p + '<img class="icon '+task.Status+'" src="/static/icons/'+task.Icon+'.svg" alt="'+task.Icon+'">'
	}
	var idx = task.Text.indexOf(" - ");
	if (idx > 0) {
		p = p + '<span class="employee">' + task.Text.substr(0,idx) + '</span>' + task.Text.substring(idx,task.Text.length);
	} else {
		p = p + task.Text;
	}
	p = p + '</p>';
	return p;
}

// ===========================================================================
// Load from server all tasks of current selected list
// ===========================================================================
function loadTasks() {
	// Purge current tasks
	$("p").remove();
	$("li").remove();
	// If no ToDo-list found - then nothing to do
	if ( !$("#task-lists-select").val() ) { return false };
	// Check selected list: if it isn't first - change current URL
	if ($("#task-lists-select").prop("selectedIndex") == 0) {
		window.history.replaceState("", "", '/');
	} else {
		window.history.replaceState("", "", '/'+$("#task-lists-select").val());
	}
	// Send Ajax POST request
	//$("#operation-status-label").text(""); // don't clear status label: for exception's of updating task or tody's-list
	showSpinner("#task-spinner-div");
	$.ajax( {
		url : "/GetTasks",
		cache: false,
		type : "post",
		dataType: "json",
		contentType: "application/json; charset=utf-8",
		data : JSON.stringify( { List: $("#task-lists-select").val()} ),
		// if success
		success: function (response) {
			hideSpinner("#task-spinner-div");
			switch(response.Result) {
  				case "SessionEmptyNotFoundOrExpired" :
  					Cookies.remove('User-Session');
  					location.reload();
					break;
				case "InvalidListName" :
					$("#operation-status-label").html(resultInvalidListName);
					break;
				case "OK" :
					// Collect tasks
					$.each(response.Tasks, function() {
						$("#"+this.Section).append(htmlPTask(this));
					});
					// Collect today's tasks
					$.each(response.TodayTasks, function() {
						if ( this.length > 0 ) {
							// insert task
							$("#today-tasks-ul").append(htmlLiTask(this));
						} else {
							// insert delimiter
							$("#today-tasks-ul").append(htmlLiDelimiter());
						}; 
					} );
					// Update Timestamps
					$("#today-tasks-ul").attr("data-timestamp", response.TodayTasksTimestamp);
					$("#tasks-main").attr("data-timestamp", response.LastModifiedTimestamp);
					// Clear status label
					$("#operation-status-label").text("");
					// Re-apply filter
					applyFilter($("#filter-input").val());
					// Calculate page statistic
					calculateStatistic();
					// Refresh events handlers
					setEvents();
					// Manage visibility and availability of list buttons
					manageListsButtons();
					// Get ready to create a new task
					$("#ib button").click();
 					break;
				default : $("#operation-status-label").html(resultUnknown);
			}
		},
		// if error returns
		error: function(jqXHR,exception) { 
			hideSpinner("#task-spinner-div");
			showAjaxError("#operation-status-label",jqXHR,exception);
		}
	} );
	return false;
}

// ===========================================================================
// Fill field of Task form values of clicked task. Prepare to task editing.
// ===========================================================================
function onTaskEdit() {
	var id
	switch ($(this).prop("tagName")) {
		case "P" :
			id = $(this).attr("id")
			break;
		case "DIV" :
			id = $(this).attr("id").substr(4);
			break;
	}
	$("#task-status-select").val( $("p#"+id).attr("class") );
	$("#task-icon-select").val( $("p#"+id).children("img").attr("alt") );
	$("#task-section-select").val( $("p#"+id).parents("section").attr("id") );
	$("#task-text-input").val( $("p#"+id).text() );
	$("#task-id-input").val( id );
	$("#task-timestamp-input").val( $("p#"+id).attr("data-timestamp") )
	if ($("div#div-"+id).length > 0) {
		$("#checkbox-today-input").prop('checked', true);
	} else {
		$("#checkbox-today-input").prop('checked', false);
	};
	$("#task-checksum-input").val( calculateTaskChecksum() );
	$("#task-submit-button").html(buttonSaveTask);
	$("#operation-status-label").text("");
	manageListsButtons();
	$("#task-text-input").focus();
}

// ===========================================================================
// Reset task form and prepare it to create new task.
// ===========================================================================
function newTask() {
	$("#task-status-select").val("created");
	$("#task-icon-select").val("");
	$("#task-section-select").val( $(this).parents("section").attr("id") );
	$("#task-text-input").val("");
	$("#task-id-input").val("");
	var now = new Date();
	$("#task-timestamp-input").val(now.toJSON());
	$("#checkbox-today-input").prop('checked', false);
	$("#task-checksum-input").val( calculateTaskChecksum() );
	$("#task-submit-button").html(buttonAddTask);
	//$("#operation-status-label").text(""); // not necessary
	manageListsButtons();
	$("#task-text-input").focus();
}

// ===========================================================================
// Calculate checksum of current task in task-form
// ===========================================================================
function calculateTaskChecksum() {
	return md5(
		$("#task-text-input").val()+
		$("#task-section-select").val()+
		$("#task-icon-select").val()+
		$("#task-status-select").val()
	);
}

// ===========================================================================
// Send current edited task (existing or new) to server and database in current list
// ===========================================================================
function submitTaskOnCurrentList() {
	// If no task text - then nothing to do
	if ( !$("#task-text-input").val() ) { 
		$("#operation-status-label").html(resultTaskEmpty); 
		$("#task-text-input").focus();
		return false
	};
	// Is any field of the task-form modified?
	if ( calculateTaskChecksum() == $("#task-checksum-input").val() ) {
		// Nothing changed
		// Update today's todo-list from the current task
		updateTodaysTasksList($("#task-id-input").val());
		// Refresh events handlers
		setEvents();
		// Get ready to create a new task
		$("#ib button").click();
		return false
	}
	// Prepare task struct
	var task = {};
	task.List = $("#task-lists-select").val();
	task.Id = $("#task-id-input").val();
	task.Text = $("#task-text-input").val().trim();
	task.Section = $("#task-section-select").val();
	task.Status = $("#task-status-select").val();
	task.Icon = $("#task-icon-select").val();
	task.Timestamp = $("#task-timestamp-input").val();
	// Send Ajax POST request
	$("#operation-status-label").text("");
	showSpinner("#task-spinner-div");
	$.ajax( {
		url : "/SendTask",
		cache: false,
		type : "post",
		dataType: "json",
		contentType: "application/json; charset=utf-8",
		data : JSON.stringify( task ),
		// if success
		success: function (response) {
			hideSpinner("#task-spinner-div");
			switch(response.Result) {
  				case "SessionEmptyNotFoundOrExpired" :
  					Cookies.remove('User-Session');
  					location.reload();
					break;
				case "TaskEmpty" :
					$("#operation-status-label").html(resultTaskEmpty);
					break;
				case "InvalidListName" :
					$("#operation-status-label").html(resultInvalidListName);
					break;
				case "UpdatedTaskNotFound" :
					$("#operation-status-label").html(resultUpdatedTaskNotFound);
					break;
				case "UpdateFailed" :
					$("#operation-status-label").html(resultUpdateFailed);
					break;
				case "InsertFailed" :
					$("#operation-status-label").html(resultInsertFailed);
					break;
				case "TaskJustUpdated" :
					$("#operation-status-label").html(resultTaskJustUpdated);
					loadTasks();
					break;
				case "TaskUpdated" :
				case "TaskInserted" :
					// Update ID and timestamp
					task.Id = response.Id;
					task.Timestamp = response.Timestamp;
					$("#tasks-main").attr("data-timestamp", response.Timestamp);
					// Add or modify task
					if (response.Result == "TaskUpdated") {
						// Update status label
						$("#operation-status-label").html(resultTaskUpdated);
						// Modify existing task
						$("#"+task.Id).replaceWith(htmlPTask(task));
						if (task.Section != $("#"+task.Id).parents("section").attr("id")) {
							// Move task <p> to another section
							$("#"+task.Id).appendTo($("#"+task.Section));
						};
					} else if (response.Result == "TaskInserted") {
						// Update status label
						$("#operation-status-label").html(resultTaskInserted);
						// Add new task to section
						$("#"+task.Section).append(htmlPTask(task));
					};
					// Update today's todo-list from the current task
					updateTodaysTasksList(task.Id);
					// Re-apply filter
					applyFilter($("#filter-input").val());
					// Calculate page statistic
					calculateStatistic();
					// Refresh events handlers
					setEvents();
					// Get ready to create a new task
					$("#ib button").click();
					break
				default : $("#operation-status-label").html(resultUnknown);
			}
		},
		// if error returns
		error: function(jqXHR,exception) { 
			hideSpinner("#task-spinner-div");
			showAjaxError("#operation-status-label",jqXHR,exception);
		}
	} );
	return false;
}

// ===========================================================================
// Move selected task to the most recent todo-list
// ===========================================================================
function moveTaskToNewList() {
	// If no task text - then nothing to do
	if ( !$("#task-text-input").val() ) { 
		$("#operation-status-label").html(resultTaskEmpty); 
		$("#task-text-input").focus();
		return false
	};
	// Prepare task struct
	var task = {};
	task.Id = $("#task-id-input").val();
	task.ToList = $("#task-lists-select").find("option:first-child").val();
	task.Text = $("#task-text-input").val().trim();
	task.Section = $("#task-section-select").val();
	task.Status = $("#task-status-select").val();
	task.Icon = $("#task-icon-select").val();
	task.Timestamp = $("#task-timestamp-input").val();
	task.Today = $("#checkbox-today-input").prop('checked');
	// Send Ajax POST request
	$("#operation-status-label").text("");
	showSpinner("#task-spinner-div");
	$.ajax( {
		url : "/MoveTask",
		cache: false,
		type : "post",
		dataType: "json",
		contentType: "application/json; charset=utf-8",
		data : JSON.stringify( task ),
		// if success
		success: function (response) {
			hideSpinner("#task-spinner-div");
			switch(response.Result) {
  				case "SessionEmptyNotFoundOrExpired" :
  					Cookies.remove('User-Session');
  					location.reload();
					break;
				case "TaskEmpty" :
					$("#operation-status-label").html(resultTaskEmpty);
					break;
				case "InvalidListName" :
					$("#operation-status-label").html(resultInvalidListName);
					break;
				case "UpdatedTaskNotFound" :
					$("#operation-status-label").html(resultUpdatedTaskNotFound);
					break;
				case "UpdateFailed" :
					$("#operation-status-label").html(resultUpdateFailed);
					break;
				case "InsertFailed" :
					$("#operation-status-label").html(resultInsertFailed);
					break;
				case "TodaysTaskListUpdateFailed" :
					$("#operation-status-label").html(resultTodaysTaskListUpdateFailed);
					break;
				case "TaskJustUpdated" :
					$("#operation-status-label").html(resultTaskJustUpdated);
					loadTasks();
					break;
				case "TaskMoved" :
					// Update timestamp
					task.Timestamp = response.Timestamp;
					$("#tasks-main").attr("data-timestamp", response.Timestamp);
					// Update status label
					$("#operation-status-label").html(resultTaskMoved);
					// Modify existing task
					$("#"+task.Id).attr("class","moved");
					$("#"+task.Id).attr("data-tooltip", statusMoved);
					$("#"+task.Id).attr("data-timestamp", response.Timestamp) ;
					$("#"+task.Id+" img").removeClass("created").removeClass("done").removeClass("canceled").addClass("moved");
					// Update today's todo-list from the current task
					if ( $("div#div-"+task.Id).length > 0 ) {
						$("div#div-"+task.Id).parents("li").replaceWith(htmlLiTask(task.Id));
					}
					// Re-apply filter
					applyFilter($("#filter-input").val());
					// Calculate page statistic
					calculateStatistic();
					// Refresh events handlers
					setEvents();
					// Get ready to create a new task
					$("#ib button").click();
					break
				default : $("#operation-status-label").html(resultUnknown);
			}
		},
		// if error returns
		error: function(jqXHR,exception) { 
			hideSpinner("#task-spinner-div");
			showAjaxError("#operation-status-label",jqXHR,exception);
		}
	} );
	return false;
}

// ===========================================================================
// 	Calculate statistics of tasks
// ===========================================================================
function calculateStatistic() {
	$("#total-tasks-count-label").text($("p").length+":");
	if ( $("p").length == 0 )
	{
		$("#done-tasks-count-label").text("0,");
		$("#canceled-tasks-count-label").text("0");
		$("#remaining-tasks-count-label").text("0");
		$("#wait-remind-tasks-count-label").text("0");
		$("#activity-tasks-count-label").text("0");
	}
	else
	{
		var total = $("p").length
		var cnt = $("p.done").length
		$("#done-tasks-count-label").text(cnt+" ("+Math.round(cnt*100/total)+"%),");
		cnt = $("p.canceled").length
		$("#canceled-tasks-count-label").text(cnt+" ("+Math.round(cnt*100/total)+"%),");
		cnt = $("img.wait, img.remind").closest("p.created").length
		$("#wait-remind-tasks-count-label").text(cnt+" ("+Math.round(cnt*100/total)+"%),");
		cnt = $("p.created").length - cnt
		$("#activity-tasks-count-label").text(cnt+" ("+Math.round(cnt*100/total)+"%)");
	}
}

// ===========================================================================
// Refresh events handlers after updating dynamics page objects
// ===========================================================================
function setEvents() {
	$("p").unbind('click');
	$("p").bind('click',onTaskEdit)
	$("div.today-task").unbind('click');
	$("div.today-task").bind('click', onTaskEdit);
	$("img.insert-delimiter").unbind('click');
	$("img.insert-delimiter").bind('click', insertDelimiter);
	$("img.delete-delimiter").unbind('click');
	$("img.delete-delimiter").bind('click', deleteDelimiter);
}

// ===========================================================================
// Restarts the timer, used to check for an update.
// ===========================================================================
function resetTimer() {
	clearTimeout(timerIdle);
	timerIdle = setTimeout(checkNeedUpdate, 5*60*1000);  // time is in milliseconds
}

// ===========================================================================
// Checks if an update tasks or today's tasks list is needed.
// ===========================================================================
function checkNeedUpdate() {
	// Send Ajax POST request
	$.ajax( {
		url : "/NeedUpdate",
		cache: false,
		type : "post",
		dataType: "json",
		contentType: "application/json; charset=utf-8",
		data : JSON.stringify( {
			List: $("#task-lists-select").val(),
			LastModifiedTimestamp: $("#tasks-main").attr("data-timestamp"),
			TodayTasksTimestamp: $("#today-tasks-ul").attr("data-timestamp")
		} ),
		// if success
		success: function (response) {
			switch(response.Result) {
  				case "SessionEmptyNotFoundOrExpired" :
  					Cookies.remove('User-Session');
  					location.reload();
					break;
				case "InvalidListName" :
					$("#operation-status-label").html(resultInvalidListName);
					break;
				case "AllActual" :
					resetTimer();
					break;
				case "NeedUpdate" :
					loadTasks();
					resetTimer();
					break;
				default : $("#operation-status-label").html(resultUnknown);
			}
		},
		// if error returns
		error: function(jqXHR,exception) { 
			hideSpinner("#task-spinner-div");
			showAjaxError("#operation-status-label",jqXHR,exception);
		}
	} );
	return false;
}

File: tasks.go
package main

import (
	"encoding/json" // to parse and generate JSON input and output parameters
	"net/http"      // for HTTP-server
	"strings"       // TrimSpace function
	"time"          // for timestamps

	"gopkg.in/mgo.v2/bson" // to use BSON queries format
)

// ===========================================================================================================================
// API: Get tasks of selected user lists.
// POST /GetTasks
// Checks the current session for validity. If the session is not valid, it returns "SessionEmptyNotFoundOrExpired" as a result.
// Returns an array of structures that identify tasks from a selected list of the current user.
// Returns LastModifiedTimestamp = max(Tasks.Timestamp)
// Also return list of today's tasks from selected list of the current user.
// Cookies: User-Session : string (UUID)
// IN: JSON: {List : string}
// OUT: JSON: { Result : string ["OK", "InvalidListName", "SessionEmptyNotFoundOrExpired"],
//				Tasks : [] { Id : string,
//							 EMail : string,
//							 List : string,
//							 Text : string,
//							 Section : string ["iu","in","nu","nn","ib"],
//							 Status : string ["created", "done", "canceled", "moved"],
//							 Icon : string ["wait","remind","call","force","mail","prepare","manage","meet","visit","make","journey","think"],
//							 Timestamp : datetime },
//				LastModifiedTimestamp : datetime,
//				TodayTasks []string (_id task or "" for delimiter),
//				TodayTasksTimestamp : datetime }
// ===========================================================================================================================

// Structure JSON-request for getting tasks
type typeGetTasksJSONRequest struct {
	List string
}

// Structure JSON-response for getting tasks
type typeGetTasksJSONResponse struct {
	Result                string
	Tasks                 []typeTask
	LastModifiedTimestamp time.Time
	TodayTasks            []string
	TodayTasksTimestamp   time.Time
}

func webGetTasks(res http.ResponseWriter, req *http.Request) {

	// Parse request to struct
	var request typeGetTasksJSONRequest
	err := json.NewDecoder(req.Body).Decode(&request)
	if err != nil {
		panic(err)
	}

	// Preparing to response
	var response typeGetTasksJSONResponse

	// Check incoming parameters
	passed, _ := TestTaskListName(request.List)
	if !passed {
		response.Result = "InvalidListName"
		ReturnJSON(res, response)
		return
	}

	// Check if the current user session is valid
	email, session := TestSession(req)
	if email == "" {
		response.Result = "SessionEmptyNotFoundOrExpired"
		ReturnJSON(res, response)
		return
	}

	// Select tasks of user list
	c := session.DB(DB).C("Tasks")
	c.Find(bson.M{"email": email, "list": request.List, "text": bson.M{"$ne": ""}}).All(&response.Tasks)

	// Select max timestamp
	var lastTask typeTask
	c.Find(bson.M{"email": email, "list": request.List, "text": bson.M{"$ne": ""}}).Sort("-timestamp").One(&lastTask)
	response.LastModifiedTimestamp = lastTask.Timestamp

	// Select today's tasks list
	var todaysTasks typeTodaysTaks
	c = session.DB(DB).C("TodaysTasks")
	c.Find(bson.M{"email": email, "list": request.List}).One(&todaysTasks)
	response.TodayTasks = todaysTasks.Tasks
	response.TodayTasksTimestamp = todaysTasks.Timestamp

	response.Result = "OK"

	// Return JSON response
	ReturnJSON(res, response)
}

// ===========================================================================================================================
// API: Update existing task from the list or append new task to the list.
// POST /SendTask
// If updated task exist in database, and its timestamp is greater than timestamp of updated task, recieved from users-application,
// return `"TaskJustUpdated"` error.
// Update existing task or generate ID and append new task to the database.
// Returns an ID and Timestamp of created or modified task.
// Cookies: User-Session : string (UUID)
// IN: JSON: { List : string,
//			   Id : string (may be null or ""),
//			   Text : string,
// 			   Section : string ["iu","in","nu","nn","ib"],
//			   Status : string ["created", "done", "canceled", "moved"],
// 			   Icon : string ["wait","remind","call","force","mail","prepare","manage","meet","visit","make","journey","think"],
//			   Timestamp : datetime (can't be null or "") }
// OUT: JSON: { Result : string ["TaskEmpty", "InvalidListName", "SessionEmptyNotFoundOrExpired", "UpdatedTaskNotFound",
//								 "UpdateFailed", "TaskJustUpdated", "TaskUpdated", "InsertFailed", "TaskInserted"],
//				Id : string,
//				Timestamp : datetime }
// ===========================================================================================================================

// Structure JSON-request for getting tasks
type typeSendTaskJSONRequest struct {
	List      string
	Id        string
	Text      string
	Section   string
	Icon      string
	Status    string
	Timestamp time.Time
}

// Structure JSON-response for getting tasks
type typeSendTaskJSONResponse struct {
	Result    string
	Id        string
	Timestamp time.Time
}

func webSendTask(res http.ResponseWriter, req *http.Request) {

	// Parse request to struct
	var request typeSendTaskJSONRequest
	err := json.NewDecoder(req.Body).Decode(&request)
	if err != nil {
		panic(err)
	}

	// Preparing to response
	var response typeSendTaskJSONResponse

	// Check incoming parameters
	request.Text = strings.TrimSpace(request.Text)
	if request.Section == "" {
		request.Section = "ib"
	}
	if request.Status == "" {
		request.Section = "created"
	}
	if request.Text == "" {
		response.Result = "TaskEmpty"
		ReturnJSON(res, response)
		return
	}
	passed, _ := TestTaskListName(request.List)
	if !passed {
		response.Result = "InvalidListName"
		ReturnJSON(res, response)
		return
	}

	// Check if the current user session is valid
	email, session := TestSession(req)
	if email == "" {
		response.Result = "SessionEmptyNotFoundOrExpired"
		ReturnJSON(res, response)
		return
	}

	// Try to update or insert task
	var task typeTask
	c := session.DB(DB).C("Tasks")
	if len(request.Id) > 0 {

		// Convert _id from string to MongoDB BSON format
		response.Id = request.Id

		// Looking for an updated task
		err = c.Find(bson.M{"email": email, "_id": bson.ObjectIdHex(response.Id)}).One(&task)
		if err != nil {
			response.Result = "UpdatedTaskNotFound"
			ReturnJSON(res, response)
			return
		}

		// Compare timestamps
		durationSeconds := task.Timestamp.Sub(request.Timestamp).Seconds()
		if durationSeconds > 0 {
			response.Result = "TaskJustUpdated"
			response.Timestamp = task.Timestamp
			ReturnJSON(res, response)
			return
		}

		// Update existing task
		response.Timestamp = time.Now().UTC()
		err = c.Update(bson.M{"email": email, "_id": task.Id},
			bson.M{"$set": bson.M{"list": request.List, "text": request.Text, "section": request.Section,
				"icon": request.Icon, "status": request.Status, "timestamp": response.Timestamp}})
		if err != nil {
			response.Result = "UpdateFailed"
			ReturnJSON(res, response)
			return
		}

		response.Result = "TaskUpdated"

	} else {

		// Create new task
		// Generate new task _id
		task.Id = bson.NewObjectId()
		task.EMail = email
		task.List = request.List
		task.Text = request.Text
		task.Section = request.Section
		task.Icon = request.Icon
		task.Status = request.Status
		task.Timestamp = time.Now().UTC()

		// Insert new task
		err = c.Insert(&task)
		if err != nil {
			response.Result = "InsertFailed"
			ReturnJSON(res, response)
			return
		}

		response.Id = task.Id.Hex()
		response.Timestamp = task.Timestamp
		response.Result = "TaskInserted"
	}

	// Return JSON response
	ReturnJSON(res, response)
}

// ===========================================================================================================================
// API: Move existing task from one list to another.
// POST /MoveTask
// Checks the current session for validity. If the session is not valid, it returns `"SessionEmptyNotFoundOrExpired"` as a result.
// If updated task exist in database, and its timestamp is greater than timestamp of updated task, recieved from users-application, return `"TaskJustUpdated"` error.
// Generate ID and append new task to destination task-list in the database.
// Update existing task in source task-list: set status = "moved".
// If Today flag is true, then add created task to today's taks list of destination task-list.
// Returns Timestamp of modified task from source task-list.
// Cookies: User-Session : string (UUID)
// IN: JSON: { Id : string (don't may be null or ""),
//			   ToList : string,
//			   Text : string,
//			   Section : string ["iu","in","nu","nn","ib"],
//			   Status : string ["created", "done", "canceled", "moved"],
//			   Icon : string ["wait","remind","call","force","mail","prepare","manage","meet","visit","make","journey","think"],
//			   Timestamp : datetime (updated task timestamp, can't be null or ""),
//			   Today : bool }
// OUT: JSON: { Result : string ["TaskEmpty", "InvalidListName", "SessionEmptyNotFoundOrExpired", "UpdatedTaskNotFound",
//								 "InsertFailed", "UpdateFailed", "TodaysTaskListUpdateFailed", "TaskJustUpdated", "TaskMoved"],
//				Timestamp : datetime }
// ===========================================================================================================================

// Structure JSON-request for getting tasks
type typeMoveTaskJSONRequest struct {
	Id        string
	ToList    string
	Text      string
	Section   string
	Icon      string
	Status    string
	Timestamp time.Time
	Today     bool
}

// Structure JSON-response for getting tasks
type typeMoveTaskJSONResponse struct {
	Result    string
	Id        string
	Timestamp time.Time
}

func webMoveTask(res http.ResponseWriter, req *http.Request) {

	// Parse request to struct
	var request typeMoveTaskJSONRequest
	err := json.NewDecoder(req.Body).Decode(&request)
	if err != nil {
		panic(err)
	}

	// Preparing to response
	var response typeMoveTaskJSONResponse

	// Check incoming parameters
	request.Text = strings.TrimSpace(request.Text)
	if request.Section == "" {
		request.Section = "ib"
	}
	if request.Status == "" {
		request.Section = "created"
	}
	if (request.Text == "") || (request.Id == "") {
		response.Result = "TaskEmpty"
		ReturnJSON(res, response)
		return
	}
	passed, _ := TestTaskListName(request.ToList)
	if !passed {
		response.Result = "InvalidListName"
		ReturnJSON(res, response)
		return
	}

	// Check if the current user session is valid
	email, session := TestSession(req)
	if email == "" {
		response.Result = "SessionEmptyNotFoundOrExpired"
		ReturnJSON(res, response)
		return
	}

	// Looking for an updated task
	var task typeTask
	c := session.DB(DB).C("Tasks")
	err = c.Find(bson.M{"email": email, "_id": bson.ObjectIdHex(request.Id)}).One(&task)
	if err != nil {
		response.Result = "UpdatedTaskNotFound"
		ReturnJSON(res, response)
		return
	}

	// Compare timestamps
	durationSeconds := task.Timestamp.Sub(request.Timestamp).Seconds()
	if durationSeconds > 0 {
		response.Result = "TaskJustUpdated"
		response.Timestamp = task.Timestamp
		ReturnJSON(res, response)
		return
	}

	// Create new task
	// Generate new task _id
	task.Id = bson.NewObjectId()
	task.EMail = email
	task.List = request.ToList
	task.Text = request.Text
	task.Section = request.Section
	task.Icon = request.Icon
	task.Status = request.Status
	task.Timestamp = time.Now().UTC()

	// Insert new task
	err = c.Insert(&task)
	if err != nil {
		response.Result = "InsertFailed"
		ReturnJSON(res, response)
		return
	}

	// Update existing task
	response.Timestamp = time.Now().UTC()
	err = c.Update(bson.M{"email": email, "_id": bson.ObjectIdHex(request.Id)},
		bson.M{"$set": bson.M{"status": "moved", "timestamp": response.Timestamp}})
	if err != nil {
		response.Result = "UpdateFailed"
		ReturnJSON(res, response)
		return
	}

	// Add the task to today's task list if needed
	if request.Today {
		// Looking for exists today's tasks list
		var todaysTasks typeTodaysTaks
		c := session.DB(DB).C("TodaysTasks")
		err = c.Find(bson.M{"email": email, "list": request.ToList}).One(&todaysTasks)
		if err == nil {
			// Update existing today's tasks list
			todaysTasks.Tasks = append(todaysTasks.Tasks, task.Id.Hex())
			todaysTasks.Timestamp = time.Now().UTC()
			err = c.Update(bson.M{"email": email, "list": request.ToList},
				bson.M{"$set": bson.M{"tasks": todaysTasks.Tasks, "timestamp": todaysTasks.Timestamp}})
			if err != nil {
				response.Result = "TodaysTaskListUpdateFailed"
				ReturnJSON(res, response)
				return
			}
			response.Result = "TodaysTaskListUpdated"
		} else {
			// Create new today's tasks list
			todaysTasks.EMail = email
			todaysTasks.List = request.ToList
			todaysTasks.Tasks = append(todaysTasks.Tasks, task.Id.Hex())
			todaysTasks.Timestamp = time.Now().UTC()
			// Insert new today's tasks list
			err = c.Insert(&todaysTasks)
			if err != nil {
				response.Result = "TodaysTaskListUpdateFailed"
				ReturnJSON(res, response)
				return
			}
		}
	}

	// Return JSON response
	response.Result = "TaskMoved"
	ReturnJSON(res, response)
}

// ===========================================================================================================================
// API: Checks need to update the task list.
// POST /NeedUpdate
// Checks the current session for validity. If the session is not valid, it returns `"SessionEmptyNotFoundOrExpired"` as a result.
// Compare LastModifiedTimestamp with max timestamp of list's tasks. If any of tasks timestamps is greater - return "NeedUpdate".
// Compare TodayTasksTimestamp with timestamp of today's task list. If today's taks list's timestamp is greater - return "NeedUpdate".
// Otherwise return "AllActual".
// Cookies: User-Session : string (UUID)
// IN: JSON: { List string,
//			   LastModifiedTimestamp : datetime,
//			   TodayTasksTimestamp : datetime }
// OUT: JSON: { Result : string ["InvalidListName", "SessionEmptyNotFoundOrExpired", "AllActual", "NeedUpdate"] }
// ===========================================================================================================================

// Structure JSON-request for getting tasks
type typeNeedUpdateJSONRequest struct {
	List                  string
	LastModifiedTimestamp time.Time
	TodayTasksTimestamp   time.Time
}

// Structure JSON-response for getting tasks
type typeNeedUpdateJSONResponse struct {
	Result string
}

func webNeedUpdate(res http.ResponseWriter, req *http.Request) {

	// Parse request to struct
	var request typeNeedUpdateJSONRequest
	err := json.NewDecoder(req.Body).Decode(&request)
	if err != nil {
		panic(err)
	}

	// Preparing to response
	var response typeNeedUpdateJSONResponse

	// Check incoming parameters
	passed, _ := TestTaskListName(request.List)
	if !passed {
		response.Result = "InvalidListName"
		ReturnJSON(res, response)
		return
	}

	// Check if the current user session is valid
	email, session := TestSession(req)
	if email == "" {
		response.Result = "SessionEmptyNotFoundOrExpired"
		ReturnJSON(res, response)
		return
	}

	// Select max timestamp
	var lastTask typeTask
	c := session.DB(DB).C("Tasks")
	c.Find(bson.M{"email": email, "list": request.List, "text": bson.M{"$ne": ""}}).Sort("-timestamp").One(&lastTask)

	// Compare timestamps
	durationSeconds := lastTask.Timestamp.Sub(request.LastModifiedTimestamp).Seconds()
	if durationSeconds > 0 {
		response.Result = "NeedUpdate"
		ReturnJSON(res, response)
		return
	}

	// Select today's tasks list
	var todaysTasks typeTodaysTaks
	c = session.DB(DB).C("TodaysTasks")
	c.Find(bson.M{"email": email, "list": request.List}).One(&todaysTasks)

	// Compare timestamps
	durationSeconds = todaysTasks.Timestamp.Sub(request.TodayTasksTimestamp).Seconds()
	if durationSeconds > 0 {
		response.Result = "NeedUpdate"
		ReturnJSON(res, response)
		return
	}

	// Return JSON response
	response.Result = "AllActual"
	ReturnJSON(res, response)
}
